SpatialGDK/Source/SpatialGDKTests/SpatialGDK/RPCService/RPCServiceTest.cpp

```diff
@@ -0,0 +1,519 @@
+
+		// Todo - check ComponentData to see if initially_present field is set.
+
+		return true;
+	}
+
+	bool PayloadWriteTest(const TArray<Worker_EntityId>& EntityIdArray, ERPCEndpointType RPCEndpointType, ERPCType RPCType, const TArray<EntityPayload>& EntityPayloads)
+	{
+		if (EntityIdArray.Num() == 0)
+		{
+			return false;
+		}
+
+		TUniquePtr<SpatialGDK::SpatialRPCService> RPCService = CreateRPCService(EntityIdArray, RPCEndpointType);
+
+		for (const EntityPayload& EntityPayloadItem : EntityPayloads)
+		{
+			if (!CheckPushRPC(*RPCService.Get(), EntityPayloadItem, RPCType, SpatialGDK::EPushRPCResult::Success))
+			{
+				return false;
+			}
+		}
+
+		TArray<SpatialGDK::SpatialRPCService::UpdateToSend> UpdatedToSendArray = RPCService->GetRPCsAndAcksToSend();
+		if (EntityPayloads.Num() != UpdatedToSendArray.Num())
+		{
+			return false;
+		}
+
+		Worker_ComponentId ExpectedUpdateComponentId = SpatialGDK::RPCRingBufferUtils::GetRingBufferComponentId(RPCType);
+
+		for (int i = 0; i < EntityPayloads.Num(); ++i)
+		{
+			const EntityPayload& EntityPayloadItem = EntityPayloads[i];
+			SpatialGDK::SpatialRPCService::UpdateToSend& Update = UpdatedToSendArray[i];
+
+			if (Update.Update.component_id != ExpectedUpdateComponentId ||
+				!CompareUpdateToSendAndEntityPayload(Update, EntityPayloadItem))
+			{
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	bool ReceiveRPCTest(const TArray<Worker_EntityId>& EntityIdArray, ERPCEndpointType RPCEndpointType, ERPCType RPCType, const EntityPayload& EntityPayloadItem)
+	{
+		USpatialStaticComponentView* StaticComponentView = NewObject<USpatialStaticComponentView>();
+		for (Worker_EntityId EntityId : EntityIdArray)
+		{
+			Worker_Authority ClientAuth = GetClientAuthorityFromRPCEndpointType(RPCEndpointType);
+			Worker_Authority ServerAuth = GetServerAuthorityFromRPCEndpointType(RPCEndpointType);
+
+			Schema_ComponentData* ClientComponentData = Schema_CreateComponentData();
+			Schema_ComponentData* ServerComponentData = Schema_CreateComponentData();
+			Schema_Object* ClientSchemaObject = Schema_GetComponentDataFields(ClientComponentData);
+			Schema_Object* ServerSchemaObject = Schema_GetComponentDataFields(ServerComponentData);
+
+			if (EntityPayloadItem.EntityId == EntityId)
+			{
+				SpatialGDK::RPCRingBufferUtils::WriteRPCToSchema(ClientSchemaObject, RPCType, 1, EntityPayloadItem.Payload);
+				SpatialGDK::RPCRingBufferUtils::WriteRPCToSchema(ServerSchemaObject, RPCType, 1, EntityPayloadItem.Payload);
+			}
+
+			AddEntityComponentToStaticComponentView(*StaticComponentView,
+				SpatialGDK::EntityComponentId(EntityId, SpatialConstants::CLIENT_ENDPOINT_COMPONENT_ID),
+				ClientComponentData,
+				ClientAuth);
+
+			AddEntityComponentToStaticComponentView(*StaticComponentView,
+				SpatialGDK::EntityComponentId(EntityId, SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID),
+				ServerComponentData,
+				ServerAuth);
+		}
+
+		bool bExtractCallbackSuccesful = true;
```

Comment:
Should this be false?

Reply:
I think true is correct. We assume it is true until it is set false. 

bool bExtractCallbackSuccesful = true;
bExtractCallbackSuccesful &= CompareRPCPayload(Payload, EntityPayloadItem.Payload);

It could be true and I could rework the logic if there is some value in that?

Reply:
Reworked this now.